libbpf-tools/tcpretrans.c

```diff
@@ -0,0 +1,359 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+/*
+ * tcpretrans  Trace IPv4 and IPv6 tcp retransmit events
+ *
+ * Copyright (c) 2020 Anton Protopopov
+ * Copyright (c) 2021 Red Hat, Inc.
+ *
+ * Based on tcpconnect.c by Anton Protopopov and
+ * tcpretrans(8) from BCC by Brendan Gregg
+ * 15-Jul-2021   Michael Gugino   Created this.
+ */
+#include <sys/resource.h>
+#include <arpa/inet.h>
+#include <argp.h>
+#include <signal.h>
+#include <limits.h>
+#include <unistd.h>
+#include <time.h>
+#include <bpf/bpf.h>
+#include "tcpretrans.h"
+#include "tcpretrans.skel.h"
+#include "trace_helpers.h"
+#include "map_helpers.h"
+
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+const char *argp_program_version = "tcpretrans 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+static const char argp_program_doc[] =
+	"\ntcpretrans: Trace TCP retransmits\n"
+	"\n"
+	"EXAMPLES:\n"
+	"    tcpretrans		# display all TCP retransmissions\n"
+	"    tcpretrans -c	# count occurred retransmits per flow\n"
+	"    tcpretrans -l	# include tail loss probe attempts\n"
+	;
+
+const char* tppath = "/sys/kernel/debug/tracing/events/tcp/tcp_retransmit_skb/id";
+
+const char* TCPSTATE[] = {
+	"ESTABLISHED",
+	"SYN_SENT",
+	"SYN_RECV",
+	"FIN_WAIT1",
+	"FIN_WAIT2",
+	"TIME_WAIT",
+	"CLOSE",
+	"CLOSE_WAIT",
+	"LAST_ACK",
+	"LISTEN",
+	"CLOSING",
+	"NEW_SYN_RECV"};
+
+
+static volatile sig_atomic_t hang_on = 1;
+
+static void sig_int(int signo)
+{
+	hang_on = 0;
+}
+
+static const struct argp_option opts[] = {
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ "count", 'c', NULL, 0, "Count connects per src ip and dst ip/port" },
+	{ "lossprobe", 'l', NULL, 0, "include tail loss probe attempts" },
+	{ "kprobe", 'k', NULL, 0, "force kprobe instead of tracepoint" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{},
+};
+
+static struct env {
+	bool verbose;
+	bool count;
+	bool lossprobe;
+	bool kprobe;
+} env = {};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'c':
+		env.count = true;
+		break;
+	case 'l':
+		env.lossprobe = true;
+		break;
+	case 'k':
+		env.kprobe = true;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void print_count_ipv4(int map_fd)
+{
+	static struct ipv4_flow_key keys[MAX_ENTRIES];
+	__u32 value_size = sizeof(__u64);
+	__u32 key_size = sizeof(keys[0]);
+	static struct ipv4_flow_key zero;
+	static __u64 counts[MAX_ENTRIES];
```

Comment:
you are mixing static and non-static variables here, that's quite confusing to read